### PR TITLE
implement routing and menu/component features for react.js

### LINKED ISSUE #11: build menu and router for react
routing: https://github.com/rackt/react-router
menu: https://github.com/instructure-react/react-menu

everything should work with foundation5.
it should be possible to hide menu items for non-logged-in users

### PR DESCRIPTION
**Pull request by benbieler**
- [x] routing
- [x] rebase master
- [x] cleanup routing
- [x] create 404 site
- [x] repair node 5.0 stuff
  - [x] fix webpack build
  - [x] repair phpstorm config
  - [x] fix tests
- [x] remove foundation, install react-bootstrap
- [x] css build fails
- [x] fix issues with vagrant
- [x] use mocha as testing engine
- [x] implement menu
- [x] fix head comments
- [x] composer update
- [x] configure frontend translation
  - [x] register component
  - [x] make language configurable through cookies
- [x] `npm test` doesn't execute all tests on linux depending on their path depth
- [x] implement security mock
  - [x] refactor cookie api
  - [x] create simple store and action mock
- [x] create abstract component
  - [x] create simple component that generates a fresh page and has a configurable menu
  - [x] implement adapter to the user system

resolves #11

### CODE REVIEW

**Comment by Ma27** [comment]
hmm, the doctrine bundle has a security vulnerability (on Sententiaregum-10 I've installed a version where that vulnerability is fixed).
I should cherry-pick that commit, so you have to rebase master (I'll notify you when you can do that)

File: package.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
   "name": "Sententiaregum",
   "version": "0.0.0",
   "description": "A social network based on Symfony2 and ReactJS",
-  "main": "index.js",
+  "main": "App.js",
```
Comment (Ma27) [inline package.json]: should be src/Frontend/App.js, shouldn't it?

File: src/Frontend/components/helloWorld.js [hunk 2]
```diff
@@ -14,5 +14,4 @@ class HelloWorld extends React.Component {
         return <h1>Hello World!</h1>;
     }
 }
-
-export default HelloWorld;
+export default HelloWorld;
```
Comment (Ma27) [inline src/Frontend/components/helloWorld.js]: please add always a newline at the end of the file

File: src/Frontend/config/routes.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+import helloWorld from '../components/helloWorld';
+import testComponent from '../components/testComponent';
+import menu from '../menu/menu';
+import Router from 'react-router';
+var DefaultRoute = Router.DefaultRoute;
+var Route = Router.Route;
+
+module.exports = (
```
Comment (Ma27) [inline src/Frontend/config/routes.js]: ecmascript6 style please (export default instead of module.exports)

File: Gruntfile.js [hunk 4]
```diff
@@ -27,7 +27,7 @@ module.exports = function (grunt) {
         },
         browserify: {
             client: {
-                src: ['src/Frontend/**/*.js', '!src/Frontend/__tests__/*.js'],
+                src: ['src/Frontend/**/*.js', '!src/Frontend/__tests__/**/*.js '],
```
Comment (Ma27) [inline Gruntfile.js]: you have to add the PathSpec _src/Frontend/App.js_ at the end of the src list in order to tell browserify to render the App.js

File: src/Frontend/App.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import React from 'react';
+import  Router from 'react-router';
```
Comment (Ma27) [inline src/Frontend/App.js]: there's one space too much

**Comment by Ma27** [comment]
@benbieler ok, I've fixed the security issue. Please rebase master

**Comment by Ma27** [comment]
btw the commit message should not be "Frontend ...", but "Task ...". The red label matters ;)

**Comment by Ma27** [comment]
but thanks a lot for the great work :+1:

**Comment by Ma27** [comment]
could you please add some tests in the end??

**Comment by benbieler** [comment]
ok

File: src/Frontend/App.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import React from 'react';
+import Router from 'react-router';
+import routes from './config/routes';
+
+Router.run(routes, function (Root) {
+    React.render(<Root />, document.getElementById('content'));
```
Comment (Ma27) [inline src/Frontend/App.js]: hmm, should we validate whether this element exist, so we should throw a proper error?

pseudocode:

``` javascript
let element = document.getElementById('content');
if (!element || element === undefined) {
    throw new Error('Cannot find page element');
}
```

File: src/Frontend/config/Routes.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ * @author Benjamin Bieler <benjaminbieler2014@gmail.com>
+ */
+import React from 'react';
+import helloWorld from '../components/HelloWorld';
+import testComponent from '../components/TestComponent';
+import menu from '../menu/view/Menu';
+import Router from 'react-router';
+
+var DefaultRoute = Router.DefaultRoute;
+var Route = Router.Route;
```
Comment (Ma27) [inline src/Frontend/config/Routes.js]: alignment

File: Vagrantfile [hunk 8]
```diff
@@ -2,14 +2,14 @@
...
-  config.vm.box = "trusty64"
+  config.vm.box = "ubuntu/trusty64"
   config.vm.synced_folder ".", "/var/www/sententiaregum", :nfs => true
 
   config.vm.hostname = "sententiaregum"
 
   config.vm.provider "virtualbox" do |vb|
     # vb settings
-    vb.name = "Sententiaregum VM"
+    vb.name = "SVM"
```
Comment (Ma27) [inline Vagrantfile]: pls remove this from the commit

File: app/SymfonyRequirements.php [hunk 9]
```diff
@@ -579,8 +579,8 @@ function_exists('simplexml_import_dom'),
 
         $this->addRecommendation(
             version_compare($installedPhpVersion, '5.4.11', '>='),
-            'When using the logout handler from the Symfony Security Component, you should have at least PHP 5.4.11 due to PHP bug #63379 (as a workaround, you can also set invalidate_session to false in the security logout handler configuration)',
-            'Install PHP 5.4.11 or newer if your project uses the logout handler from the Symfony Security Component.'
+            'When using the logout handler from the Symfony security Component, you should have at least PHP 5.4.11 due to PHP bug #63379 (as a workaround, you can also set invalidate_session to false in the security logout handler configuration)',
```
Comment (Ma27) [inline app/SymfonyRequirements.php]: this, too

File: src/Frontend/config/MenuConfig.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * @author Benjamin Bieler <benjaminbieler2014@gmail.com>
+ */
+
+'use strict';
+
+import MenuItem from '../menu/model/MenuItem';
+let config = [
+    new MenuItem('/#!/', 'Label', 'User')
```
Comment (Ma27) [inline src/Frontend/config/MenuConfig.js]: the _/#!/_ prefix should be added automatically by the MenuItem class

File: src/Frontend/menu/model/MenuItem.js [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * @author Benjamin Bieler <benjaminbieler2014@gmail.com>
+ */
+class MenuConfig {
+    label;
+    url;
+    permissionRule;
+
+    constructor(label, url, permissionRule) {
+        this.label = label;
```
Comment (Ma27) [inline src/Frontend/menu/model/MenuItem.js]: alignment

File: src/Frontend/menu/view/Menu.js [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+ * @copyright (c) 2015 Sententiaregum
+ * Please check out the license file in the document root of this application
+ */
+import React from 'react';
+import MenuConfig from '../model/MenuItem.js';
+/**
+ * @author Benjamin Bieler <benjaminbieler2014@gmail.com>
+ */
+class Menu extends React.Component {
+    render() {
```
Comment (Ma27) [inline src/Frontend/menu/view/Menu.js]: could you please add some jsdoc stuff to your methods?

File: src/Frontend/security/Security.js [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+ * Sententiaregum is a social network based on Symfony2 and ReactJS
+ *
+ * @copyright (c) 2015 Sententiaregum
+ * Please check out the license file in the document root of this application
+ */
+
+/**
+ * @author Benjamin Bieler <benjaminbieler2014@gmail.com>
+ */
+class security {
```
Comment (Ma27) [inline src/Frontend/security/Security.js]: Should be uppercase I guess ;)

File: Vagrantfile [hunk 14]
```diff
@@ -2,7 +2,7 @@
 # vi: set ft=ruby :
 
 Vagrant.configure(2) do |config|
-  config.vm.box = "trusty64"
+  config.vm.box = "ubuntu/trusty64"
```
Comment (Ma27) [inline Vagrantfile]: this should be reverted, too as this will be fixed in another PR. (I guess, your machine works fine, so don't need the vagrantfile anymore)

File: src/Frontend/config/Routes.js [hunk 15]
```diff
@@ -0,0 +1,24 @@
+/*
+ * This file is part of the sententiaregum application.
+ *
+ * Sententiaregum is a social network based on Symfony2 and ReactJS
+ *
+ * @copyright (c) 2015 Sententiaregum
+ * Please check out the license file in the document root of this application
+ */
+
+/**
```
Comment (Ma27) [inline src/Frontend/config/Routes.js]: jsdoc should be used at classes only

File: src/Frontend/menu/view/Menu.js [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+                menuItems.push(
+                    <li className={api.getMenuClass(items[i].url)} key={i}>
+                        <a href={items[i].url}>
+                            {SenTranslation.trans(items[i].label, {})}
+                        </a>
+                    </li>
+                );
+            }
+        }
+        return (
```
Comment (Ma27) [inline src/Frontend/menu/view/Menu.js]: some spaces would be nice, this looks very squeezed ;)

File: src/Frontend/security/Security.js (hunk 13 above)
Comment (Ma27) [inline src/Frontend/security/Security.js]: hmm, I would refactor the whole security system using reflux:
we need a UserStore that is able to get the securityexpression from the menuconfig in order to evaluate it. After that an event will be triggered using the reflux dispatcher (we DON'T return true because this won't work when using ajax requests in order to check roles for example).
The event should be a class from the Frontend/event directory and should contain information about the user and whether he's allowed to see the menu item.
In your case you just need to build the basic class and trigger **always** an event that tells the menu that the user is allowed to see every menuitem

File: src/Frontend/menu/model/MenuItem.js [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+ * Sententiaregum is a social network based on Symfony2 and ReactJS
+ *
+ * @copyright (c) 2015 Sententiaregum
+ * Please check out the license file in the document root of this application
+ */
+
+/**
+ * @author Benjamin Bieler <benjaminbieler2014@gmail.com>
+ */
+class MenuConfig {
```
Comment (Ma27) [inline src/Frontend/menu/model/MenuItem.js]: this should be moved to a global model directory.
*Frontend/model/menu/MenuConfig.js*

File: src/Frontend/menu/view/Menu.js [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * @copyright (c) 2015 Sententiaregum
+ * Please check out the license file in the document root of this application
+ */
+import React from 'react';
+import MenuConfig from '../model/MenuItem.js';
+/**
+ * @author Benjamin Bieler <benjaminbieler2014@gmail.com>
+ */
+class Menu extends React.Component {
```
Comment (Ma27) [inline src/Frontend/menu/view/Menu.js]: should be moved to *Frontend/components/menu/Menu.js*

**Comment by Ma27** [comment]
rebased this

**Comment by Ma27** [comment]
resolved all conflicts with master, so we can continue working on that branch @benbieler

**Comment by Ma27** [comment]
it seems as node :shit: 5 has some issues with our build :/

**Comment by Ma27** [comment]
ok it was some weird configuration issue. Should do now.

**Comment by Ma27** [comment]
ah karma does not work because of those weird peer deps

File: src/Frontend/styles/custom.scss [hunk 19]
```diff
@@ -8,7 +8,7 @@
  * file that was distributed with this source code.
  */
 
-@import '../../../node_modules/zurb-foundation-5/scss/foundation';
+$icon-font-path:'~bootstrap-sass/assets/fonts/bootstrap/';
```
Comment (Ma27) [inline src/Frontend/styles/custom.scss]: obsolete

File: webpack.factory.js [hunk 20]
```diff
@@ -11,28 +11,47 @@
...
             comments: /^\\/
         })); // quick and dirty fix that doesn't match any comments
 
-        cssQueryString = 'minimize&sourceMap';
+        useMinimizeAndSourceMap = 'minimize&sourceMap';
     }
 
-    cssQueryString = 'css-loader?' + cssQueryString + '!sass-loader?outputStyle=expanded&sourceMap';
+    scssQueryString = 'css-loader?' + useMinimizeAndSourceMap + '!sass-loader?outputStyle=expanded&sourceMap';
+    cssQueryString  = 'css-loader?' + useMinimizeAndSourceMap;
```
Comment (Ma27) [inline webpack.factory.js]: can be done better

**Comment by Ma27** [comment]
ok now we're using less instead of scss in order to avoid bigger issues with executables

File: src/AppBundle/Composer/ScriptHandler.php [hunk 21]
```diff
@@ -38,6 +38,14 @@ public static function installNpmDependencies(CommandEvent $event)
...
         $process = new Process(sprintf('%s install --no-bin-links', $npm), null, null, null, 1000);
         $process->run($handler);
+    }
+
+    public static function buildFrontendData(CommandEvent $event)
+    {
+        $npm     = (new ExecutableFinder())->find('npm');
+        $handler = function ($type, $buffer) use ($event) {
+            $event->getIO()->write($buffer, false);
+        };
```
Comment (Ma27) [inline src/AppBundle/Composer/ScriptHandler.php]: should be refactored

**Comment by Ma27** [comment]
can be useful for translations: https://github.com/martinandert/react-translate-component

File: src/Frontend/components/app/Menu.js [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+ * @author Maximilian Bosch <maximilian.bosch.27@gmail.com>
+ */
+class Menu extends ReactComponent {
+    /**
+     * Builds a virtual dom for the menu section.
+     *
+     * @returns {XML}
+     */
+    render() {
+        let items = this.props.items.map(function (item) {
```
Comment (Ma27) [inline src/Frontend/components/app/Menu.js]: es6 style pls

**Comment by Ma27** [comment]
/cc @benbieler 
I guess it is better to use mocha:
sice mocha works properly with the nodejs core, so we don't have any problems with a non-working debugger and big sourcemap issues. Another reason is that mocha supports real commonjs support and since all of the view is abstracted in ReactJS components (there's an integration of reactjs and its testutils into mocha) and the rest is pure business logic, the switch to a node-based engine (karma was browser-based) shouldn't be a hard job
